### ISSUE TITLE
Slack direct messages not working

### ISSUE THREAD

**Issue by pastorhudson** [skoczen/will#369]
It seems sending to a user ID in slack doesn't work like I used to?
If I try:
`self.say("Test message", channel = self.message.data.sender.id)`
I get back:
`Fri, 20 Jul 2018 23:43:29 [ERROR] I was asked to post to the slack U08FPSYTW channel, but it doesn't exist.`
When I send a direct message to the bot and:
`print(self.message.data.channel.id)`
It prints `DBH45HA59`. I can't figure out how to get `DBH45HA59` from the user `U08FPSYTW`.
@Ashex Any ideas on how to send DM through the slack backend?

**Comment by pastorhudson** [comment]
If I look in 
`self.load('slack_channel_cache', {})`
This is what's stored for DM channels.
```
This is the key: DBH45HA59
This is the content: {'id': 'DBH45HA59', 'name': 'DBH45HA59', 'source': {'id': 'DBH45HA59', 'members': [], 'name': 'DBH45HA59'}, 'members': {}}
```
If  someone could point me toward where this is built I could perhaps figure out how to get name to be stored as U08FPSYTW

**Comment by Ashex** [comment]
You want to pass the message content back with `say` so that Will knows it's a private message and handles it appropriately.

**Comment by pastorhudson** [comment]
It didn't work.         
`will.say('"%s" channel added to the whitelist.' % will.message.data.channel.name.title(), message=will.message)
`
Sends to the same channel the command was recieved from. Unless I'm missing something?

**Comment by Ashex** [comment]
Ah so you want to take a message from a channel and act on it by sending a direct message, that's a bit different. 

You're sending a message directly to a user so you want to actually look inside `slack_people_cache` to get their user ID for sending messages (unless it's a group chat).

I haven't actually done what you're trying to accomplish but it's on my list of enhancements to make. The message should already include their user ID so you either want to override portions of the message dictionary in the response or add a new function that will send direct messages.

**Comment by pastorhudson** [comment]
Right! So I tried that and I get a user ID 'U08FPSYTW' and I try to send a message to that user. channel='U08FPSYTW'  and I get Mon, 23 Jul 2018 08:20:56 [ERROR] I was asked to post to the slack U08FPSYTW channel, but it doesn't exist. When I look at the slack api it seems the DM channel has it's own name seperate from the user name. I can't find that DM channel ID anywhere in slack_people_cache. The way to get it is to make an api call to `"im.open", user="U08FPSYTW"`
This returns something like this:
```
{
    "ok": true,
    "no_op": true,
    "already_open": true,
    "channel": {
        "id": "D19U4244U"
    }
}
```
So where should we add this call? Is this something we should cache in slack_people.cache? Or is a better way to have channel try an im.open when it can't find a channel?

**Comment by pastorhudson** [comment]
I found a fix that works for me. I can do a PR if the solution is acceptable.
I added:
    ```
def get_im(self, user_id):
        return self.client.api_call(
            "im.open",
            user=user_id)['channel']['id']
```

And added "channel" to Person:
```
people[k] = Person(
                id=v.id,
                mention_handle="<@%s>" % v.id,
                handle=v.name,
                source=clean_for_pickling(v),
                name=v.real_name,
                channel=self.get_im(v.id)
            )
```

I called it channel because that's the first thing I tried. You can now DM anybody using:
`say("Some DM message", channel=self.message.data.sener.channel)`
Channel is also populated with the DM channel in slack_people_cache.

Is this an acceptable solution?

**Comment by pastorhudson** [comment]
Another option would be to put it in the say function where it would catch a slack user ID inputted as a channel and convert it to the IM channel. But I like having all the dm channels in slack_people_cache

**Comment by Ashex** [comment]
I'm just a contributor so it's ultimately up to @skoczen how this should be integrated. I'm a bit worried about storing the DM channels in the cache as this can become resource intensive on large slack workspaces and potentially cause startup delays while it makes the `im.open` call for each user at it populates the cache. We can reduce the impact with a condition where it only adds/updates where the channel key is missing.

I like the second option better, perhaps added an argument to the say function that would explicitly call it a direct user message but the user ID always starts with `U` so that's probably superfluous.

**Comment by pastorhudson** [comment]
That makes a lot of sense. I'll work it up and make a pull request. @skoczen can take a look at the pr.

**Comment by pastorhudson** [comment]
Submitted solution here #375

**Comment by pastorhudson** [comment]
I'm not sure if this is needed anymore. I'm merging in the new slack update and I'll report back.

**Comment by Ashex** [comment]
I had to do some extra condition checks in the slack backend for channel checks and also set `is_direct=True` so it would know how to process the channel lookup. With the slack update it can reply to DMs but unsolicited (i.e responding to a message in a channel by sending a DM) doesn't work for some reason even though logging shows a success so I was going to dive deeper as I did sort it out in my code base about 6 months ago. 

I opened #429 for that specifically so if you sort it with your PR we can close that out.

**Comment by pastorhudson** [comment]
I just got contract for another project that will be consuming cycles for a couple weeks. So if you have the time go for it!
